python/tk_framework_desktopserver/server_protocol.py

```diff
@@ -123,45 +126,81 @@ def onMessage(self, payload, isBinary):
+
+        # Run each request from a thread, even though it might be something very simple like opening a file. This
         # will ensure the server is as responsive as possible. Twisted will take care of the thread.
-        reactor.callInThread(self._process_message, message_host, message)
+        reactor.callInThread(
+            self._process_message,
+            message_host,
+            message,
+            message["protocol_version"],
+        )
 
-    def _process_message(self, message_host, message):
+    def _process_message(self, message_host, message, protocol_version):
 
         # Retrieve command from message
         command = message["command"]
 
         # Retrieve command data from message
-        data = {}
-        if "data" in command:
-            data = command["data"]
-
+        data = command.get("data", dict())
         cmd_name = command["name"]
 
         # Create API for this message
         try:
             # Do not resolve to simply ShotgunAPI in the imports, this allows tests to mock errors
-            shotgun = shotgun_api.ShotgunAPI(message_host, self.process_manager)
+            api = get_shotgun_api(
+                protocol_version,
+                message_host,
+                self.process_manager,
+                wss_key=self._wss_key,
+            )
         except Exception, e:
-            message_host.report_error("Error in loading ShotgunAPI. " + e.message)
+            message_host.report_error("Unable to get a ShotgunAPI object: %s" % e)
             return
 
         # Make sure the command is in the public API
-        if cmd_name in shotgun.public_api:
+        if cmd_name in api.PUBLIC_API_METHODS:
             # Call matching shotgun command
             try:
-                func = getattr(shotgun, cmd_name)
-                func(data)
+                func = getattr(api, cmd_name)
+                requires_sync = (cmd_name in api.SYNCHRONOUS_METHODS)
             except Exception, e:
-                message_host.report_error("Error! Could not execute function (possibly wrong command arguments) for [%s] -- %s"
-                                          % (cmd_name, e.message))
+                message_host.report_error(
+                    "Could not find API method %s: %s" % (cmd_name, e)
+                )
+            else:
+                # If a method is expecting to be run synchronously we
+                # need to make sure that happens. An example of this is
+                # the get_actions method in v2 of the api, which might
+                # trigger a cache update. If that happens, we can't let
+                # multiple cache processes occur at once, because each
+                # is bootstrapping sgtk, which won't play well if multiple
+                # are occurring at the same time, all of which potentially
+                # copying/downloading files to disk in the same location.
+                if requires_sync:
```

Comment:
It's for a couple of reasons:

1. In many cases, we have a single row in the cache that can be used my most get_actions requests by the client. This is especially true for tk-config-basic setups. As such, it makes sense to allow first-in-first-out behavior, as many (sometimes all) subsequent get_actions calls will then be able to return cached data. If they're all running async at roughly the same time, we'll likely trigger multiple cache inserts/updates due to race conditions.
2. The sqlite database that acts as the data store for the cache can only handle a single writing connection at a time. We don't want a bunch of caching subprocesses to be kicked off async at the same time, causing lots of sqlite inserts/updates at the same time.